commit 0095d641fe5bd80e61222f0c7bfc4b437006e161 (refs/changes/99/4699/4)
Author: Dave Eddy <dave@daveeddy.com>
Date:   2018-08-20T17:13:46-04:00 (1 year, 2 months ago)
    
    joyent/node-manta#251 specifying a path multiple times to mfind results in crash
    Reviewed by: Robert Mustacchi <rm@joyent.com>

diff --git a/bin/mfind b/bin/mfind
index 820582b..8f65083 100755
--- a/bin/mfind
+++ b/bin/mfind
@@ -8,6 +8,7 @@ var http = require('http');
 var https = require('https');
 var path = require('path');
 var url = require('url');
+var util = require('util');
 
 var assert = require('assert-plus');
 var bunyan = require('bunyan');
@@ -156,8 +157,9 @@ function printEntry(opts, obj) {
         }
     });
 
-    options.paths.forEach(function (p) {
-        barrier.start(p);
+    options.paths.forEach(function (p, i) {
+        var id = util.format('%d %s', i, p);
+        barrier.start(id);
         client.ftw(p, options, function (err, res) {
             if (err) {
                 if (err.name === 'InvalidDirectoryError') {
@@ -168,14 +170,14 @@ function printEntry(opts, obj) {
                     if (options.limit && totalEntries >= options.limit) {
                         process.exit(0);
                     }
-                    barrier.done(p);
+                    barrier.done(id);
                     return;
                 } else if (err.name === 'NotFoundError') {
                     if (err.path === undefined)
                         err.path = p;
                     printError(err.path, err);
                     lastError = err;
-                    barrier.done(p);
+                    barrier.done(id);
                     return;
                 } else {
                     var epath = p;
@@ -187,7 +189,7 @@ function printEntry(opts, obj) {
             }
 
             res.on('entry', print);
-            res.on('end', barrier.done.bind(barrier, p));
+            res.on('end', barrier.done.bind(barrier, id));
         });
     });
 })();
diff --git a/test/mfind.test.js b/test/mfind.test.js
index bc77413..28165a6 100644
--- a/test/mfind.test.js
+++ b/test/mfind.test.js
@@ -123,6 +123,19 @@ test('mfind TESTDIR', function (t) {
     });
 });
 
+/*
+ * joyent/node-manta#251 specifying a path multiple times to mfind results in
+ * crash.
+ */
+test('mfind TESTDIR TESTDIR (same argument multiple times)', function (t) {
+    forkExecWait({
+        argv: [MFIND, TESTDIR, TESTDIR]
+    }, function (err, info) {
+        t.ifError(err, err);
+        t.done();
+    });
+});
+
 test('mfind -j TESTDIR', function (t) {
     forkExecWait({
         argv: [MFIND, '-j', TESTDIR]
